### PR TITLE
feat: make sure a player can save replays [CU-2jzdzq3]

### PR DESCRIPTION
**Pull request by bwpikaard**

### CODE REVIEW

**Comment by 0c370t** [comment]
Task linked: [CU-2jzdzq3 Make Sure Someone can Save Replays](https://app.clickup.com/t/2jzdzq3)

File: core/src/internal/franchise/player/player.service.ts [hunk 1]
```diff
@@ -424,4 +424,26 @@ export class PlayerService {
             };
         }
     }
+
+    async playerCanSaveDemos(playerId: number): Promise<boolean> {
+        const player = await this.playerRepository.getById(playerId, {
+            relations: {
+                member: {
+                    platformAccounts: true,
```
Comment (0c370t) [inline core/src/internal/franchise/player/player.service.ts:432]: Is there a way to filter these into the correct game automatically? Feels like the steps beneath could be done at the database to do less work.

**Review by 0c370t — APPROVED** [review]

File: core/src/internal/scrim/scrim.mod.resolver.ts [hunk 2]
```diff
@@ -159,6 +160,7 @@ export class ScrimModuleResolver {
         @Args("leaveAfter", {type: () => Int}) leaveAfter: number,
         @Args("group", {nullable: true}) groupKey?: string,
         @Args("createGroup", {nullable: true}) createGroup?: boolean,
+        @Args("canSaveDemos", {type: () => Boolean, nullable: true}) canSaveDemos?: boolean,
```
Comment (zachstence) [inline core/src/internal/scrim/scrim.mod.resolver.ts:163]: Why is this an arg to the mutation? Shouldn't we get this info from the implicitly player making the request?

File: microservices/matchmaking-service/src/scrim/scrim-logic/scrim-logic.service.ts [hunk 3]
```diff
@@ -25,24 +27,28 @@ export class ScrimLogicService {
...
+            this.analyticsService
+                .send(AnalyticsEndpoint.Analytics, {
+                    name: "scrimPopped",
+                    strings: [["scrimId", scrim.id]],
+                })
+                .catch(err => {
+                    this.logger.error(err);
+                });
+        } else {
+            await this.scrimService.cancelScrim(scrim.id);
```
Comment (zachstence) [inline microservices/matchmaking-service/src/scrim/scrim-logic/scrim-logic.service.ts:50]: We should probably do something other than blindly cancelling. Can we send a notification to players saying why the scrim was cancelled?

**Review by zachstence — CHANGES_REQUESTED** [review]
Instead of cancelling the scrim if nobody can save replays, I think it would be better to prevent somebody from joining if it would fill the scrim with people that can't save replays. I.e., there's 3 console players in a 2s scrim, when a 4th console player tries to join, it doesn't let them, but the other 3 remain in the scrim.

Thoughts?

File: core/src/internal/scrim/scrim.mod.resolver.ts (hunk 2 above)
Comment (bwpikaard) [reply core/src/internal/scrim/scrim.mod.resolver.ts:163]: Unsure. Ticket said to add a checkbox for the player that they can save it that will only show up if they can save replays. Use case might be they have accounts that can save demos, but they're playing on one that can't. @0c370t

File: microservices/matchmaking-service/src/scrim/scrim-logic/scrim-logic.service.ts (hunk 3 above)
Comment (bwpikaard) [reply microservices/matchmaking-service/src/scrim/scrim-logic/scrim-logic.service.ts:50]: Not through UI unless we add a "cancelled reason" to the scrim cancelled event or if we just do a Discord DM, which we should do but that shouldn't be THE source.

File: core/src/internal/franchise/player/player.service.ts (hunk 1 above)
Comment (bwpikaard) [reply core/src/internal/franchise/player/player.service.ts]: #485f8e1

**Review by zachstence — APPROVED** [review]

**Pull request merged**